openshift/odo
# add container name validation during creation and push

Signed-off-by: Stephanie <yangcao@redhat.com>

**What type of PR is this?**
/kind bug
/area devfile



**What does does this PR do / why we need it**:
Add  validation for devfile container name during component creation, and push. Block the action if the container name is not a valid K8s resourece name. 

**Which issue(s) this PR fixes**:

Fixes #3698 

**PR acceptance criteria**:




- [ ] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

test component creation:
create a devfile with a long container name, and run `odo create --devfile <devfile path>`
```
$ odo create --devfile ./devfile.yaml 
Validation
 ✓  Creating a devfile component from devfile path: /Users/stephanie/Documents/devfile-spring2/devfile.yaml [62590ns]
 ✓  Validating devfile component [277583ns]
 ✗  endpoint name "runtimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntime" is not valid, endpoint name should conform the following requirements: 
- Contain at most 63 characters
- Contain only lowercase alphanumeric characters or ‘-’
- Start with an alphanumeric character
- End with an alphanumeric character
- Must not contain all numeric values
```

test odo push:
`odo create nodejs --starter`
modify the devfile to use a long container name
run `odo push`
```
$ odo push
 ✗  Validating K8s Resources Name [132166ns]
 ✗  endpoint name "runtimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntimeruntime" is not valid, endpoint name should conform the following requirements: 
- Contain at most 63 characters
- Contain only lowercase alphanumeric characters or ‘-’
- Start with an alphanumeric character
- End with an alphanumeric character
- Must not contain all numeric values

```
